src/main/cpp/RobotContainer.cpp

```diff
@@ -4,18 +4,131 @@
 
 #include "RobotContainer.h"
 
-RobotContainer::RobotContainer() : m_autonomousCommand(&m_subsystem) {
-  // Initialize all of your commands and subsystems here
+#include <frc/smartdashboard/SmartDashboard.h>
 
-  // Configure the button bindings
-  ConfigureButtonBindings();
+RobotContainer::RobotContainer()
+{
+    // Initialize the hardware.
+    VOKC_CALL(this->InitHardware());
+
+    // Initialize subsystems.
+    VOKC_CALL(this->InitDrivetrain());
+
+    // Configure the button bindings
+    ConfigureButtonBindings();
+}
+
+void RobotContainer::ConfigureButtonBindings()
+{
+    // Configure your button bindings here
+}
+
+frc2::Command *RobotContainer::GetAutonomousCommand()
+{
+    // An example command will be run in autonomous
+    return &m_autonomousCommand;
+}
+
+bool RobotContainer::InitHardware()
+{
+    // Initialize the hardware interface.
+    hardware_ = std::make_unique<HardwareInterface>();
+
+    // Initialize sub-hardware interfaces.
+    hardware_->actuators = std::make_unique<ActuatorInterface>();
+    hardware_->sensors = std::make_unique<SensorInterface>();
+
+    // Initialize the actuators.
+    OKC_CALL(this->InitActuators(hardware_->actuators.get()));
+
+    // Set up sensors.
+    OKC_CALL(
+        this->InitSensors(*hardware_->actuators, hardware_->sensors.get()));
+
+    return true;
+}
+
+bool RobotContainer::InitActuators(ActuatorInterface *actuators_interface)
+{
+    OKC_CHECK(actuators_interface != nullptr);
+
+    // Initialize drivetrain motors.
+    actuators_interface->left_motor_1 =
```

Comment:
hol' up I thought hardware was initialized and all that inside the subsystem? How is the subsystem supposed to access the hardware if it's being managed inside the --well, I suppose that's what the "actuators_interface" is for. Seems like it would be annoying to initialize *all* the hardware in the RobotContainer class as opposed to the individual subsystems.

Reply:
The idea is to init everything once in RobotContainer and provide access to the hardware using subsystem interfaces and pointers. That way we can easily verify our hardware configuration in a centralized location and we can have all the checks in one place